File: packages/sdk/java/packages/admin/src/main/java/dev/edgebase/sdk/admin/AdminAuthClient.java
// EdgeBase Java SDK — Admin authentication client.
// Server-side user management with Service Key authentication.
package dev.edgebase.sdk.admin;

import dev.edgebase.sdk.core.*;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Admin auth client for server-side user management.
 *
 * <p>
 * Only available via {@link AdminEdgeBase} /
 * {@link EdgeBase#server(String, String)}.
 *
 * <p>
 * Usage:
 * 
 * <pre>{@code
 * var admin = EdgeBase.admin("https://...", "sk-...");
 * var user = admin.adminAuth().getUser("user-id");
 * admin.adminAuth().setCustomClaims("user-id", Map.of("role", "pro"));
 * }</pre>
 */
public class AdminAuthClient {
    private final HttpClient client;
    private final String serviceKey;

    public AdminAuthClient(HttpClient client, String serviceKey) {
        this.client = client;
        this.serviceKey = serviceKey;
    }

    private void requireServiceKey() {
        if (serviceKey == null || serviceKey.isEmpty()) {
            throw new EdgeBaseError(403,
                    "Service Key required for admin operations. Use EdgeBase.admin(url, serviceKey).");
        }
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getUser(String userId) {
        requireServiceKey();
        return (Map<String, Object>) client.get("/auth/admin/users/" + userId);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> listUsers() {
        return listUsers(null, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> listUsers(Integer limit, String cursor) {
        requireServiceKey();
        Map<String, String> params = new HashMap<>();
        if (limit != null)
            params.put("limit", limit.toString());
        if (cursor != null)
            params.put("cursor", cursor);
        return (Map<String, Object>) client.get("/auth/admin/users", params.isEmpty() ? null : params);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> createUser(Map<String, Object> data) {
        requireServiceKey();
        return (Map<String, Object>) client.post("/auth/admin/users", data);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> updateUser(String userId, Map<String, Object> data) {
        requireServiceKey();
        return (Map<String, Object>) client.patch("/auth/admin/users/" + userId, data);
    }

    public void deleteUser(String userId) {
        requireServiceKey();
        client.delete("/auth/admin/users/" + userId);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> setCustomClaims(String userId, Map<String, Object> claims) {
        requireServiceKey();
        return (Map<String, Object>) client.put("/auth/admin/users/" + userId + "/claims", claims);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> revokeAllSessions(String userId) {
        requireServiceKey();
        return (Map<String, Object>) client.post("/auth/admin/users/" + userId + "/revoke", Collections.emptyMap());
    }
}

File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/AdminAuthClient.kt
// EdgeBase Kotlin SDK — Admin authentication client.
//
// Server-side user management with Service Key authentication.
//
// Usage:
// ```kotlin
// val client = EdgeBase.server("https://...", serviceKey = System.getenv("EDGEBASE_SERVICE_KEY") ?: "")
// val user = admin.adminAuth.getUser("user-id")
// admin.adminAuth.setCustomClaims("user-id", mapOf("role" to "pro"))
// ```

package dev.edgebase.sdk.admin

import dev.edgebase.sdk.admin.generated.GeneratedAdminApi
import dev.edgebase.sdk.core.*

/**
 * Admin auth client for server-side user management.
 *
 * Only available via [ServerEdgeBase] / [EdgeBase.server()].
 * Service Key is injected automatically via [HttpClient].
 *
 * All HTTP calls delegate to [GeneratedAdminApi] — no hardcoded API paths.
 */
class AdminAuthClient(
    private val client: HttpClient,
    private val serviceKey: String?,
    private val adminCore: GeneratedAdminApi
) {
    private fun requireServiceKey() {
        if (serviceKey.isNullOrEmpty()) {
            throw EdgeBaseError(
                statusCode = 403,
                message = "Service Key required for admin operations. Use EdgeBase.server(url, serviceKey)."
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun getUser(userId: String): Map<String, Any?> {
        requireServiceKey()
        val response = adminCore.adminAuthGetUser(userId) as Map<String, Any?>
        // Server returns { user: { id, email, ... } } — unwrap to flat map for callers.
        @Suppress("UNCHECKED_CAST")
        return (response["user"] as? Map<String, Any?>) ?: response
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun listUsers(limit: Int? = null, cursor: String? = null): Map<String, Any?> {
        requireServiceKey()
        val params = buildMap<String, String> {
            if (limit != null) put("limit", limit.toString())
            if (cursor != null) put("cursor", cursor)
        }
        return adminCore.adminAuthListUsers(params.ifEmpty { null }) as Map<String, Any?>
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun createUser(data: Map<String, Any>): Map<String, Any?> {
        requireServiceKey()
        return adminCore.adminAuthCreateUser(data) as Map<String, Any?>
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun updateUser(userId: String, data: Map<String, Any>): Map<String, Any?> {
        requireServiceKey()
        return adminCore.adminAuthUpdateUser(userId, data) as Map<String, Any?>
    }

    suspend fun deleteUser(userId: String) {
        requireServiceKey()
        adminCore.adminAuthDeleteUser(userId)
    }

    /**
     * Set custom JWT claims for a user.
     *
     * Delegates to [GeneratedAdminApi.adminAuthSetClaims] — PUT /api/auth/admin/users/{id}/claims.
     *
     * ```kotlin
     * admin.adminAuth.setCustomClaims("user-id", mapOf("role" to "pro", "tier" to "premium"))
     * ```
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun setCustomClaims(userId: String, claims: Map<String, Any?>): Map<String, Any?> {
        requireServiceKey()
        return adminCore.adminAuthSetClaims(userId, claims) as Map<String, Any?>
    }

    /**
     * Revoke all sessions for a user.
     *
     * Forces the user to sign in again on all devices.
     * Delegates to [GeneratedAdminApi.adminAuthRevokeUserSessions].
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun revokeAllSessions(userId: String): Map<String, Any?> {
        requireServiceKey()
        return adminCore.adminAuthRevokeUserSessions(userId) as Map<String, Any?>
    }
}
